### ISSUE TITLE
2.0: Swift Replicator.addChangeListener crashes

### ISSUE THREAD

**Issue by jayzhanghs** [couchbase/couchbase-lite-ios#1926]
let replicator = Replicator(config: replicatorConfig)
        replicator.addChangeListener { (replicatorChange) in
        }

In my project the above code crashes at addChangeListener even if I have only an empty block, with errors "fatal error: attempted to read an unowned reference but object 0x60c000129560 was already deallocated". Really suspect that the `unowned self` designator in the framework definition of `addChangeListener` may have caused this. 

Commenting out this call would stop the crash, but I also lost capability to track the changes from replicator.

--------
* Version:   2.0DB016
* Client OS: iOS 11
* Server:    Sync Server
* Environment: Xcode 9, Swift 3.2

**Comment by pasin** [comment]
@rajagp are you seeing the same thing is the app you are working on?

**Comment by rajagp** [comment]
I am not seeing any crashes.  This is what I have. 
```
       _pushPullRepl = Replicator.init(config: config)
_pushPullReplListener = _pushPullRepl?.addChangeListener({ [weak self] (change) in
            let s = change.status
            print("PushPull Replicator: \(s.progress.completed)/\(s.progress.total), error: \(String(describing: s.error)), activity = \(s.activity)")
            if s.progress.completed == s.progress.total {
                self?.postNotificationOnReplicationState(.idle)
            }
            else {
                self?.postNotificationOnReplicationState(s.activity)
            }
        })
        

```
@jayzhanghs : Can you change the scope of the replicator object that you are creating so it has class scope and it is not locally scoped to the function.
For instance,in my case `fileprivate var _pushPullRepl:Replicator?` is defined as a file private variable scoped to my class.

**Comment by jayzhanghs** [comment]
@rajagp Yes, changing replicator object as an instance variable did make it work! 

In the sample given in CouchLite 2.0 the replicator was created as a local variable. Please update [the document](https://developer.couchbase.com/documentation/mobile/2.0/guides/couchbase-lite/native-api/replication/index.html?language=swift).

Thanks!

**Comment by rajagp** [comment]
Filed a ticket https://github.com/couchbaselabs/couchbase-mobile-portal/issues/816

**Comment by pasin** [comment]
@jayzhanghs please make the instance variable for the replicator. In Objective-C version, the started replicator is cached inside the database object as well so that you don't need to make the instance variable. I had the changes for this but I would like to hold it off a bit as there is a bit more work needs to be done to take care the case when closing the database. To sum up, the fix will not be in the next DB release which is DB18.

**Comment by pasin** [comment]
Look like the fix for this issue is not enough for a non-continuous replicator. I'm reopening the issue.

**Comment by pasin** [comment]
Reference: CBSE-5304

**Comment by pasin** [comment]
This issue is not trivial to fix as the notification could be posted later in a queue. The workaround would be retaining the replicator object in an instance variable.

**Comment by pasin** [comment]
After reviewing this issue again, the fix was fairy simple. The fix will be available in 2.1.
